docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,317 @@
+|>Write Blob Stream                  |  9,1976|  0,0000|    218|      0|      0|      0|    457|     46|
+|>Fixup Metadata                     |  9,2033|  0,0002|    218|      0|      0|      0|    457|     46|
+|>Generated IL and metadata          |  9,2102|  0,0015|    218|      0|      0|      0|    457|     46|
+|>Layout image                       |  9,2195|  0,0025|    218|      0|      0|      0|    457|     46|
+|>Writing Image                      |  9,2272|  0,0006|    218|      0|      0|      0|    456|     46|
+|>Signing Image                      |  9,2338|  0,0000|    218|      0|      0|      0|    456|     46|
+// main6
+|Write .NET Binary                   |  9,2463|  0,1930|    218|      0|      0|      0|    464|     46|     <-- long
+--------------------------------------------------------------------------------------------------------
+```
+
+The phases taking the longest are marked in the right. Those are assembly import, type check, optimization and IL writing.
+
+## Implementation plan
+
+The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and harder to review. Therefore, I propose to implement the feature in stages.
+
+### What should be done
+
+**Stage 1 - omit retypecheck for plain rebuild**
+
+That is, after running "fsc test.fs", rerunning `fsc test.fs` will omit type-checking step. This should work for multiple files (project) as well. Everything will be happening in `main1` - after parsing, we check the cache and deserialize the typecheck info if available, passing it to `main2` right away. If not, we do the normal type check, caching the results at the end of it, prior to passing them to the next stage. This will speed up scenarios like tinkering with plain "fsc.exe" or project rebuild when the results are incorrectly invalidated.
+
+**Stage 2 - minimize retypecheck in slightly modified projects**
+
+That is, when rebuilding the project after modifying an independent file, only the modified file will be retypechecked, for others the check will be omitted. This will handle many target scenarios for the feature, like modifying one test in a test project.
+
+**Stage 3 - omit reresolving assemblies**
+
+That is, for either of the scenarios above, skip the `Import mscorlib+FSharp.Core` step. This logic is also done in `main1`, prior to parsing. Same ideas apply - if cached, deserialize the information - otherwise, follow the normal process and then cache the results. The data should be stored separately from the typecheck info, so that the information be reused partially.
+
+**Stage 4 - omit reoptimization**
+
+That is, for everything that doesn't need to be retypechecked, don't do reoptimization either. This happens in the `main3` phase of compilation. 
```

Comment:
This makes little sense. With small changes we will need to run optimisation either way. With no changes, nothing will be rebuilt. Already isn't, there are up to date checks in place already.

Reply:
I would go even further - do not bother with opt phase at all.
This is about improving inner dev loop by making it fast - turning off optimizations (e.g. by using DEBUG) is something users should be already doing.

I think the rare occasions of debugging an optimization-related issue will be fine with running the full optimization phase.

Reply:
Yeah, that makes even more sense. I think we should focus on debug config scenarios, as it's the most common one for the dev loop. And in debug we don't optimize.

Reply:
Alright, removed for brevity and clarity